

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>genewiseMAFC &mdash; ALISEQ 1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> ALISEQ
          

          
            
            <img src="../_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../welcome.html">ALISEQ, a program for analysing RNA-Seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../setup.html">Instalation and setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Running the software</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow.html">Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../more.html">Other stuff</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules.html">Python modules</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ALISEQ</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="index.html">Module code</a> &raquo;</li>
        
      <li>genewiseMAFC</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for genewiseMAFC</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="sd">.. Created on Wed Aug 29 12:37:31 2018</span>

<span class="sd">   @author: cow082</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">sys</span>


<div class="viewcode-block" id="meanAbsoluteFoldChange"><a class="viewcode-back" href="../genewiseMAFC.html#genewiseMAFC.meanAbsoluteFoldChange">[docs]</a><span class="k">def</span> <span class="nf">meanAbsoluteFoldChange</span><span class="p">(</span><span class="n">frame</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;find the average log2 fold change of SNPs (alleles) for a single gene</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#print(&#39;meanAbsoluteFoldChange: run...&#39;, frame)</span>
   
    <span class="n">frame</span><span class="p">[</span><span class="s1">&#39;fc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">frame</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="o">.</span><span class="n">a_altCount</span> <span class="o">/</span> <span class="n">row</span><span class="o">.</span><span class="n">a_refCount</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    
    <span class="n">frame</span><span class="p">[</span><span class="s1">&#39;log2fc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">frame</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">fc</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    
    <span class="n">frame</span><span class="p">[</span><span class="s1">&#39;absLog2fc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">frame</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="nb">abs</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">log2fc</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    
    <span class="n">frame</span><span class="p">[</span><span class="s1">&#39;aveLog2fc&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">frame</span><span class="p">[</span><span class="s1">&#39;absLog2fc&#39;</span><span class="p">])</span>
    
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">({</span>
            <span class="s1">&#39;snp_count&#39;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">frame</span><span class="p">)),</span>
            <span class="s1">&#39;ave_read_depth&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">frame</span><span class="o">.</span><span class="n">a_altCount</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">frame</span><span class="o">.</span><span class="n">a_refCount</span><span class="p">))</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">frame</span><span class="p">),</span> <span class="mi">1</span><span class="p">),</span>
            <span class="s1">&#39;aveLog2fc&#39;</span><span class="p">:</span> <span class="n">frame</span><span class="p">[</span><span class="s1">&#39;aveLog2fc&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="p">})</span></div>

    
<div class="viewcode-block" id="enrichASE"><a class="viewcode-back" href="../genewiseMAFC.html#genewiseMAFC.enrichASE">[docs]</a><span class="k">def</span> <span class="nf">enrichASE</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;identify genes with differential expression between two alleles</span>
<span class="sd">    </span>
<span class="sd">    requires: annotated SNP data (*.merged.csv) output from bedtools intersect</span>
<span class="sd">    </span>
<span class="sd">    How it works:</span>
<span class="sd">        1. group SNPs together by gene</span>
<span class="sd">        2. For each SNP in that gene, calculate the absolute(log2(fold change))</span>
<span class="sd">        3. take the average of absolute(log2(fold change)) values across all SNPS for that gene</span>
<span class="sd">        4. filter on:</span>
<span class="sd">            - average read depth at variable positions for the gene</span>
<span class="sd">            - number of SNPs in the gene</span>
<span class="sd">            - average(absolute(log2(fold change))) for the gene</span>
<span class="sd">    </span>
<span class="sd">    Two optional args with no defaults change the program behaviour:</span>
<span class="sd">        1. --output &lt;file&gt; sends the results to file rather than stdout</span>
<span class="sd">        2. --gene &lt;gene&gt; yields data for every SNP in a specific gene. Without</span>
<span class="sd">           this arg, the program produces instead a summary table of all </span>
<span class="sd">           significant genes in order of priority </span>
<span class="sd">           (sorted by: average(absolute(log2(fold change)))) </span>
<span class="sd">           </span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># the input has no header, so supply one</span>
    <span class="n">headers</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;a_chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;a_chromStart&#39;</span><span class="p">,</span> <span class="s1">&#39;a_chromEnd&#39;</span><span class="p">,</span> <span class="s1">&#39;a_refBase&#39;</span><span class="p">,</span> 
               <span class="s1">&#39;a_altBase&#39;</span><span class="p">,</span> <span class="s1">&#39;a_refCount&#39;</span><span class="p">,</span> <span class="s1">&#39;a_altCount&#39;</span><span class="p">,</span> 
               <span class="s1">&#39;b_chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;b_chromStart&#39;</span><span class="p">,</span> <span class="s1">&#39;b_chromEnd&#39;</span><span class="p">,</span> <span class="s1">&#39;b_featureType&#39;</span><span class="p">,</span> 
               <span class="s1">&#39;b_gene&#39;</span><span class="p">,</span> <span class="s1">&#39;b_orientation&#39;</span><span class="p">]</span>
    
    <span class="c1"># read the input data</span>
    <span class="n">df1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">input</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">headers</span><span class="p">)</span>
    
    <span class="c1"># calculate meanAbsoluteFoldChange per gene</span>
    <span class="n">df2</span> <span class="o">=</span> <span class="n">df1</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;b_gene&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">meanAbsoluteFoldChange</span><span class="p">)</span>
    
    <span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="o">.</span><span class="n">sort_values</span><span class="p">([</span><span class="s1">&#39;aveLog2fc&#39;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># filter hits</span>
    <span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="p">[</span><span class="n">df2</span><span class="p">[</span><span class="s1">&#39;ave_read_depth&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">args</span><span class="o">.</span><span class="n">minCoverage</span><span class="p">]</span>
    <span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="p">[</span><span class="n">df2</span><span class="p">[</span><span class="s1">&#39;snp_count&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">args</span><span class="o">.</span><span class="n">minSnps</span><span class="p">]</span>
    <span class="n">df2</span> <span class="o">=</span> <span class="n">df2</span><span class="p">[</span><span class="n">df2</span><span class="p">[</span><span class="s1">&#39;aveLog2fc&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">math</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">minFoldChange</span><span class="p">,</span> <span class="mi">2</span><span class="p">)]</span>
    
    <span class="n">df2</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;gene&#39;</span>
    
    <span class="c1"># filter the main data table to identify expression at the per-SNP level </span>
    <span class="c1"># for the significant genes identified above</span>
    <span class="n">hits</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">({</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">df2</span><span class="o">.</span><span class="n">index</span><span class="p">)})</span>
    <span class="n">df3</span> <span class="o">=</span> <span class="n">df1</span><span class="p">[</span><span class="n">df1</span><span class="o">.</span><span class="n">b_gene</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">hits</span><span class="o">.</span><span class="n">gene</span><span class="p">)]</span>
    <span class="n">df3</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;b_gene&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">df3</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;gene&#39;</span>
    <span class="n">df3</span> <span class="o">=</span> <span class="n">df3</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="s1">&#39;b_chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;b_chromStart&#39;</span><span class="p">,</span> <span class="s1">&#39;b_chromEnd&#39;</span><span class="p">,</span> 
                    <span class="s1">&#39;b_featureType&#39;</span><span class="p">,</span> <span class="s1">&#39;b_orientation&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">df3</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;refBase&#39;</span><span class="p">,</span> 
                   <span class="s1">&#39;altBase&#39;</span><span class="p">,</span> <span class="s1">&#39;refCount&#39;</span><span class="p">,</span> <span class="s1">&#39;altCount&#39;</span><span class="p">]</span>
    
    <span class="c1"># choose where to send the output</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">:</span>
        <span class="n">save_location</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">output</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">save_location</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">stdout</span>
        
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gene</span><span class="p">:</span>
        <span class="c1"># print specific details for a single target gene</span>
        <span class="n">df3</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">gene</span><span class="p">,:]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">save_location</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># display a summary of all hits</span>
        <span class="n">df2</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df2</span><span class="p">))</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">save_location</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span></div>


<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-i&#39;</span><span class="p">,</span> <span class="s1">&#39;--input&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;annotated SNP data from bedtools intersect&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="s1">&#39;--minCoverage&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> 
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;minimun average read depth at variable sites&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="s1">&#39;--minSnps&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> 
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;minimum number of SNPs in a given gene&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-f&#39;</span><span class="p">,</span> <span class="s1">&#39;--minFoldChange&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;minimum average absolute fold change&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="s1">&#39;--gene&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> 
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;gene of interest&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="s1">&#39;--output&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> 
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;output file&quot;</span><span class="p">)</span>
    <span class="n">enrichASE</span><span class="p">(</span><span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">())</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>

</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Chris Cowled

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../_static/jquery.js"></script>
        <script type="text/javascript" src="../_static/underscore.js"></script>
        <script type="text/javascript" src="../_static/doctools.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    

  

  <script type="text/javascript" src="../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>